internal/datasources/project_events.go

```diff
@@ -0,0 +1,226 @@
+package datasources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/api"
+	"github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/errors"
+	providerschema "github.com/couchbasecloud/terraform-provider-couchbase-capella/internal/schema"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &ProjectEvents{}
+	_ datasource.DataSourceWithConfigure = &ProjectEvents{}
+)
+
+// ProjectEvents is the ProjectEvents data source implementation.
+type ProjectEvents struct {
+	*providerschema.Data
+}
+
+// NewProjectEvents is a helper function to simplify the provider implementation.
+func NewProjectEvents() datasource.DataSource {
+	return &ProjectEvents{}
+}
+
+// Metadata returns the backup data source type name.
+func (d *ProjectEvents) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_project_events"
+}
+
+// Schema defines the schema for the ProjectEvents data source.
+func (d *ProjectEvents) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = ProjectEventsSchema()
+}
+
+// Read refreshes the Terraform state with the latest data of ProjectEvents.
+func (d *ProjectEvents) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state providerschema.ProjectEvents
+	diags := req.Config.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var (
+		organizationId = state.OrganizationId.ValueString()
+		projectId      = state.ProjectId.ValueString()
+		clusterIds     []string
+		userIds        []string
+		severityLevels []string
+		tags           []string
+		from           string
+		to             string
+		page           int
+		perPage        int
+		sortBy         string
+		sortDir        string
+		err            error
+	)
+	queryParam := make(map[string][]string)
+
+	if !state.ClusterIds.IsNull() && !state.ClusterIds.IsUnknown() {
```

Comment:
Nit: Can we have a separate function call like `Validate` for these null and unknown checks?